### PR TITLE
ffbidx package for the fast feedback indexer library

### PR DESCRIPTION
**Pull request by hcstadler**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/ffbidx/meta.yaml
Comment (ndevenish) [inline recipes/ffbidx/meta.yaml:27]: You don't need `conda-forge::` here

Comment (hcstadler) [reply recipes/ffbidx/meta.yaml]: Done

**Comment by hcstadler** [comment]
@conda-forge/help-c-cpp, ready for review

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [inline recipes/ffbidx/meta.yaml:5]: Please indent with two spaces like we do in all other recipes.

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [inline recipes/ffbidx/meta.yaml:24]: This should be in host

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [inline recipes/ffbidx/meta.yaml:29]: These four should also be in host

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [inline recipes/ffbidx/meta.yaml:35]: This should no longer be used 
```suggestion
```

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [inline recipes/ffbidx/meta.yaml:37]:
```suggestion
test:
  requires:
    - pip
  commands:
    - pip check
```

File: recipes/ffbidx/build.sh
Comment (xhochy) [inline recipes/ffbidx/build.sh:1]: If you move the dependencies into host, you should no longer need this
```suggestion
```

File: recipes/ffbidx/build.sh
Comment (xhochy) [inline recipes/ffbidx/build.sh:2]:
```suggestion
CUDA_HOME=${BUILD_PREFIX} meson setup --reconfigure --buildtype=release --prefix=${PREFIX} --libdir=lib -Ddefault_library=shared -Dinclude-python-api=enabled meson $MESON_ARGS
```

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/ffbidx/meta.yaml
Comment (hcstadler) [reply recipes/ffbidx/meta.yaml:5]: Done

File: recipes/ffbidx/meta.yaml
Comment (hcstadler) [reply recipes/ffbidx/meta.yaml:24]: Done

File: recipes/ffbidx/meta.yaml
Comment (hcstadler) [reply recipes/ffbidx/meta.yaml:29]: Done.

File: recipes/ffbidx/meta.yaml
Comment (hcstadler) [reply recipes/ffbidx/meta.yaml:35]: Done

File: recipes/ffbidx/build.sh
Comment (hcstadler) [reply recipes/ffbidx/build.sh:1]: Done

File: recipes/ffbidx/build.sh
Comment (hcstadler) [reply recipes/ffbidx/build.sh:2]: Done

File: recipes/ffbidx/meta.yaml
Comment (hcstadler) [reply recipes/ffbidx/meta.yaml:37]: Done

**Comment by hcstadler** [comment]
This fails on cuda_118 (which I thought I had skipped) in "Packaging cbfmt" - something I do not use for my code.

File: recipes/ffbidx/meta.yaml
Comment (ndevenish) [reply recipes/ffbidx/meta.yaml]: Are you sure about this? Don't compilers normally go in build? (because they don't need to be tied to the target architecture, and otherwise would prevent e.g. cross compilation).

Or did you just mean cuda-cudart-dev? (which I agree looks like it should be in host)

File: recipes/ffbidx/meta.yaml
Comment (ndevenish) [inline recipes/ffbidx/meta.yaml:25]: Usually I would expect these four should be in `build:`. But I don't know if meson needs that to workaround the environment.

Comment (ndevenish) [reply recipes/ffbidx/meta.yaml:25]: Ah, meson was reviewed as needing to be in host, above, so maybe it does belong. It probably doesn't matter that much, if it doesn't export a run requirement.

**Comment by hcstadler** [comment]
The linter fails if compiler('cxx') and stdlib('c') are not present in build.
cuda_120 fails if compiler('cuda') is not present in build.

**Comment by hcstadler** [comment]
@conda-forge/help-c-cpp, ready for review

File: recipes/ffbidx/meta.yaml
Comment (ndevenish) [reply recipes/ffbidx/meta.yaml]: I am pretty sure that this was a mobile-editing error, and that it was only intended to say that cuda-cudart-dev should be in host. So the compilers and `stdlib("c")` should be in build, not host.

**Comment by ndevenish** [comment]
- I'm pretty sure compilers should definitely not be in `host:`, otherwise I don't think cross-compilation would work. I think above request to move them to host was using an incorrect line selection.
- pkg-config shouldn't _need_ to be in host, because that's just the CLI tool, and meson should still be able to pick up the other dependencies in host when the actual `pkg-config` executable is on the PATH outside the host path. I am not excluding the possibility that meson does something weird requiring it though, so (unless review disagrees) it's probably fine to have in host, along with meson (I don't think either declare an associated run requirement).
- The top two `cuda_home` lines in build.sh look as though they aren't doing anything?

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [reply recipes/ffbidx/meta.yaml:27]: It will matter for cross-compilation probably and with having meson here, you avoid having a python installation in build.

**Comment by hcstadler** [comment]
@ndevenish: I agree with you on that.
The cuda_home lines in build.sh compute the cuda home path (for a typical installation) from the location of the nvcc binary, which in a typical installation is under bin/nvcc of the cuda home path. If I don't set CUDA_HOME, the meson build fails as far as I remember.
Much of my and mesons confusion stems from the fact that compiler('cuda') results in fundamentally different cuda installations for cuda_118 and cuda_120. I decided to only support cuda_120.

File: recipes/ffbidx/meta.yaml
Comment (xhochy) [reply recipes/ffbidx/meta.yaml:27]: I think this got mixed up by the GitHub app. Compilers should only be in build.

**Comment by dagewa** [comment]
This came up in our DIALS core meeting today. I'm a bit confused about the status, are we waiting for re-review by @xhochy?

**Comment by xhochy** [comment]
This us waiting for the removal of the compilers from host

File: recipes/ffbidx/meta.yaml
Comment (carterbox) [inline recipes/ffbidx/meta.yaml:30]:
```suggestion
requirements:
  build:
    - {{ compiler('cuda') }}
    - {{ compiler('cxx') }}
    - {{ stdlib('c') }}
    - meson
    - pkg-config
  host:
    - eigen  # header only library
    - python
    - numpy
```

File: recipes/ffbidx/meta.yaml
Comment (carterbox) [inline recipes/ffbidx/meta.yaml:15]:
```suggestion
  skip: True  # [cuda_compiler_version in (None, "None", "11.8")]
```

File: recipes/ffbidx/build.sh
Comment (carterbox) [inline recipes/ffbidx/build.sh:3]: Looks like you neeed to add some args to help meson find NumPy in the host prefix.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Review by xhochy — APPROVED** [review]

**Review by carterbox — CHANGES_REQUESTED** [review]
If you look at the logs, you will see that the python versions in the build ($BUILD_PREFIX) and host ($PREFIX) environment are different and that meson detects the wrong one (the one in build instead of host). This is why I asked you to add some extra args to your build script to help meson find the python/numpy from the host prefix ($PREFIX).

**Comment by hcstadler** [comment]
@carterbox: Thanks for the great review and suggestions. I added an option to select the python installation in my meson build and updated the recipe, so the python installation to build against is now set to the one installed under $PREFIX.

**Comment by hcstadler** [comment]
@conda-forge/help-c-cpp, ready for review

**Comment by xhochy** [comment]
Had to merge main here as the build was building a different recipe than the one in the PR.

**Review by xhochy — APPROVED** [review]

**Pull request merged**
